### DISCUSSION TITLE
Multiple foreign keys in secondary table

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10866] [Usage Questions] [answered]
Hi everyone. I need help with SQLAlchemy, or rather with two ForeignKeys in a Many-To-Many bundle table.
I have such a sign:
```py
a_b = Table(
"a_b",
    Base.metadata,
    Column("a_id", ForeignKey("a.id"), primary_key=True),
    Column("b_id", ForeignKey("b.id"), primary_key=True),
    Column("custom_a_id", ForeignKey("a.id "), primary_key=True, nullable=True)
)
```
In the `a_b` table I have the relations `a` and `b`, but there is also a `custom_a_id` - this is a, which can optionally be linked to `b`.
`A` and `B` themselves are below:
```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[uuidpk]
    ...
```
```py
class B(Base):
    __tablename__ = "b"

    id: Mapped[uuidpk]
    ...
    a: Mapped[list["a"]] = relationship(
        "a",
        secondary=a_b,
        lazy="joined",
        foreign_keys="[a_b.c.a_id]"
    )
```
What I want: 
If in the `a_b` table, `custom_a_id` is not NULL, then `B.a = A(id=custom_a_id)`.
If in the `a_b` table, `custom_a_id` is NULL, then `B.a = A(id=a_id)`.

Is it really possible to do this? And if possible, please tell me how? I've tried a lot of things, but it doesn't work out

**Comment by klaydev** [comment]
If you are interested, the NoForeignKeysError error occurs

**Comment by zzzeek** [comment, marked as answer]
sure, you have to set up `primaryjoin` and `secondaryjoin` to appropriate SQL expressions that select for these columns as you want.  the SQL COALESCE function is useful here.

```py
uuidpk = Annotated[uuid.UUID, mapped_column(primary_key=True)]

Base = declarative_base()

a_b = Table(
    "a_b",
    Base.metadata,
    Column("a_id", ForeignKey("a.id"), primary_key=True),
    Column("b_id", ForeignKey("b.id"), primary_key=True),
    Column(
        "custom_a_id", ForeignKey("a.id"), primary_key=True, nullable=True
    ),
)


class A(Base):
    __tablename__ = "a"

    id: Mapped[uuidpk]


class B(Base):
    __tablename__ = "b"

    id: Mapped[uuidpk]
    ...
    a: Mapped[list[A]] = relationship(
        secondary=a_b,
        # just to make the SQL more readable for the moment
        # lazy="joined",
        secondaryjoin=lambda: foreign(func.coalesce(a_b.c.custom_a_id, a_b.c.a_id))
        == A.id,
    )


print(
    select(B).join(B.a)
)
```

the SQL (without the joinedload for the moment) is

```
SELECT b.id 
FROM b JOIN a_b AS a_b_1 ON b.id = a_b_1.b_id JOIN a ON coalesce(a_b_1.custom_a_id, a_b_1.a_id) = a.id

```

coalesce returns the first expression that's not NULL.

**Reply by klaydev** [reply]
Thank you very much! You've saved me a lot of time!
